### PR TITLE
add keccak hash

### LINKED ISSUE #30: keccak

### PR DESCRIPTION
**Pull request by AntoineFONDEUR**
add keccak hash

Time spent on this PR: 0.5 days

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Resolves #30

### CODE REVIEW

File: src/builtins/keccak/keccak.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+//https://github.com/XKCP/XKCP/blob/master/Standalone/CompactFIPS202/Python/CompactFIPS202.py
+
+function ROL64(a: bigint, n: number): bigint {
+  return (
+    ((a >> (BigInt(64) - BigInt(n % 64))) + (a << BigInt(n % 64))) %
+    (BigInt(1) << BigInt(64))
+  );
+}
+
+function KeccakF1600onLanes(lanes: bigint[][]): bigint[][] {
```
Comment (ClementWalter) [inline src/builtins/keccak/keccak.ts:11]:
```suggestion
function keccakF1600onLanes(lanes: bigint[][]): bigint[][] {
```

File: src/builtins/keccak/keccak.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
+//This is the typescript version of the compact python implementation by the Keccak team :
+//https://github.com/XKCP/XKCP/blob/master/Standalone/CompactFIPS202/Python/CompactFIPS202.py
+
+function ROL64(a: bigint, n: number): bigint {
```
Comment (ClementWalter) [inline src/builtins/keccak/keccak.ts:4]:
```suggestion
function rol64(a: bigint, n: number): bigint {
```

File: src/builtins/keccak/keccak.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+  let lanes = new Array(5)
+    .fill(0)
+    .map((_, x) =>
+      new Array(5)
+        .fill(BigInt(0))
+        .map((_, y) =>
+          load64(state.slice(8 * (x + 5 * y), 8 * (x + 5 * y) + 8)),
+        ),
+    );
+  lanes = KeccakF1600onLanes(lanes);
```
Comment (ClementWalter) [inline src/builtins/keccak/keccak.ts:78]:
```suggestion
  lanes = keccakF1600onLanes(lanes);
```

File: src/builtins/keccak/keccak.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+  const newState = new Uint8Array(200);
+  for (let x = 0; x < 5; x++) {
+    for (let y = 0; y < 5; y++) {
+      newState.set(store64(lanes[x][y]), 8 * (x + 5 * y));
+    }
+  }
+  return newState;
+}
+
+function Keccak(
```
Comment (ClementWalter) [inline src/builtins/keccak/keccak.ts:88]:
```suggestion
function keccak(
```

File: src/builtins/keccak/keccak.ts [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+  let inputOffset = 0;
+  while (inputOffset < inputBytes.length) {
+    blockSize = Math.min(inputBytes.length - inputOffset, rateInBytes);
+    for (let i = 0; i < blockSize; i++) {
+      state[i] ^= inputBytes[i + inputOffset];
+    }
+    inputOffset += blockSize;
+    if (blockSize === rateInBytes) {
+      state = KeccakF1600(state);
```
Comment (ClementWalter) [inline src/builtins/keccak/keccak.ts:112]:
```suggestion
      state = keccakF1600(state);
```

File: src/builtins/keccak/keccak.ts [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+function store64(a: bigint): Uint8Array {
+  const b = new Uint8Array(8);
+  for (let i = 0; i < 8; i++) {
+    b[i] = Number((a >> BigInt(8 * i)) % BigInt(256));
+  }
+  return b;
+}
+
+function KeccakF1600(state: Uint8Array): Uint8Array {
```
Comment (ClementWalter) [inline src/builtins/keccak/keccak.ts:68]:
```suggestion
function keccakF1600(state: Uint8Array): Uint8Array {
```

File: src/builtins/keccak/keccak.ts [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+    inputOffset += blockSize;
+    if (blockSize === rateInBytes) {
+      state = KeccakF1600(state);
+      blockSize = 0;
+    }
+  }
+
+  state[blockSize] ^= delimitedSuffix;
+  if ((delimitedSuffix & 0x80) !== 0 && blockSize === rateInBytes - 1) {
+    state = KeccakF1600(state);
```
Comment (ClementWalter) [inline src/builtins/keccak/keccak.ts:119]:
```suggestion
    state = keccakF1600(state);
```

File: src/builtins/keccak/keccak.ts [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+  }
+  state[rateInBytes - 1] ^= 0x80;
+  state = KeccakF1600(state);
```
Comment (ClementWalter) [inline src/builtins/keccak/keccak.ts:122]:
```suggestion
  state = keccakF1600(state);
```

File: src/ui/Code.ts [hunk 9]
```diff
@@ -142,3 +142,18 @@ function CHECK_ECDSA_SIGNATURE(
 function RANGE_CHECK(num: number): bigint {
   return range_check(BigInt(num));
 }
+
+function KECCAK(message: number | string): number | string {
+  const utf8Bytes = encodeUTF8(message.toString());
+  const bytearrayOutput = Keccak(1088, 512, utf8Bytes, 0x01, 256 / 8);
```
Comment (ClementWalter) [inline src/ui/Code.ts:148]:
```suggestion
  const bytearrayOutput = keccak(1088, 512, utf8Bytes, 0x01, 256 / 8);
```

File: src/ui/Code.ts [hunk 10]
```diff
@@ -142,3 +142,18 @@ function CHECK_ECDSA_SIGNATURE(
 function RANGE_CHECK(num: number): bigint {
   return range_check(BigInt(num));
 }
+
+function KECCAK(message: number | string): number | string {
```
Comment (ClementWalter) [inline src/ui/Code.ts:146]:
```suggestion
function KECCAK(message: number | string): string {
```

File: src/ui/Code.ts [hunk 11]
```diff
@@ -142,3 +142,18 @@ function CHECK_ECDSA_SIGNATURE(
...
+
+  const bitOutput = Array.from(bytearrayOutput)
+    .map((byte) => byte.toString(2).padStart(8, "0"))
+    .join("");
+
+  const bitStringSlice = bitOutput.slice(6);
+  const bitBigInt = BigInt("0b" + bitStringSlice);
+
+  const hexOutput = "0x" + bitBigInt.toString(16);
+  return hexOutput;
```
Comment (ClementWalter) [inline src/ui/Code.ts:158]: I don't understand this

Comment (AntoineFONDEUR) [reply src/ui/Code.ts:158]: Starknet keccak is the 250 first bits of the regular keccak-256, so this converts keccak output to bits, selects the 250 first and converts it to hex

File: src/ui/Code.ts [hunk 12]
```diff
@@ -139,10 +139,25 @@ function RANGE_CHECK(num: number | string): string {
...
+
+function KECCAK(message: number | string): string {
+  const utf8Bytes = encodeUTF8(message.toString());
+  const bytearrayOutput = keccak(1088, 512, utf8Bytes, 0x01, 256 / 8);
+
+  const bitOutput = Array.from(bytearrayOutput)
+    .map((byte) => byte.toString(2).padStart(8, "0"))
+    .join("");
+
+  const bitStringSlice = bitOutput.slice(6);
```
Comment (ClementWalter) [inline src/ui/Code.ts:154]:
```suggestion
  // trim the first 6 bits as Starknet keccak is keccak % 2**250
  const bitStringSlice = bitOutput.slice(6);
```

File: src/ui/Code.ts [hunk 13]
```diff
@@ -139,10 +139,26 @@ function RANGE_CHECK(num: number | string): string {
...
 
-function POSEIDON(x: number | string, y: number | string): number | string {
-  return "0x" + poseidon(BigInt(x), BigInt(y)).toString(16);
-}
-
 function PEDERSEN(x: number | string, y: number | string): number | string {
   return "0x" + pedersen(BigInt(x), BigInt(y)).toString(16);
 }
+
+function KECCAK(message: number | string): string {
```
Comment (ClementWalter) [inline src/ui/Code.ts:146]:
```suggestion
function KECCAK(message: string): string {
```

**Review by ClementWalter — APPROVED** [review]

**Pull request merged**